azavea/cac-tripplanner
# Feature/kak/admin tours#1108

## Overview

Adds the ability to create tours that are collections of destinations in the admin interface. Each destination has the additional properties of a required order field and optional start and end dates/times.

Allows ordering destinations in the admin interface by setting the 'order' field.

As with destinations and events, tours have a 'published' property so that unpublished tours can be saved, but hidden from display.


### Demo

'Tours' is a new option under 'destinations and events':
![image](https://user-images.githubusercontent.com/960264/63555471-0c1e3b80-c50f-11e9-9869-373851dd560c.png)

Start and end times for the destinations are optional.

![image](https://user-images.githubusercontent.com/960264/63555499-2b1ccd80-c50f-11e9-8357-6ba2e8b22c1d.png)

Validation checks that the destinations for the tour are unique, as are the orderings, and that the end date/time is not earlier than the start date/time:

![image](https://user-images.githubusercontent.com/960264/63555549-5b646c00-c50f-11e9-96ff-876d965276a4.png)


### Notes

Clicking the button to add another destination defaults to the new destination having an 'order' of 1. It would be nice to autoincrement the 'order' for the user, but doing so would require overriding some jQuery deep in the Django templates, so I didn't pursue that.


## Testing Instructions

 * Run migrations
 * Go to http://localhost:8024/admin/destinations/tour/
 * Create a tour
 * Check validation ensures destinations and orders within a tour are unique
 * Check validation ensures end time is not before start time
 * Create a second tour; ensure validation allows it to have destinations and/or orders that are the same as the first tour (uniqueness check is only within a tour)
 * When editing an existing tour, the new destination inline form should have a default 'order' value that matches its row number
 * When editing an existing tour, the destinations inline forms should be listed ordered by their 'order' value by default


Closes #1108
